Add a content libraries import from courseware API and management command [SE-4619]

jvdm opened the pull request:
## Description

This PR introduces the import from course functionality to content libraries.

The functionality is exposed by (a.) A RESTful API, and (b.) A management command. Both are added to Content Libraries V2. They allow block importing from a local modulestore instance. Additionally, the command line supports importing blocks from a remote platform instance through API calls.

The management command allows platform operators to import blocks from existing courseware and remote instances into blockstore to allow them to be accessed through Libraries v2.

The APIs allow course teams to manage background tasks (Celery workers) that will perform the importing while providing persisted information for querying the state of the tasks. It is suitable to be used by [UI interfaces to allow customers of Libraries V2](https://github.com/edx/frontend-app-library-authoring/pull/24) to select Organisations and Courses to import content from and query the state and progress of the import task.

Adding this command and API doesn't cause any immediate implication to users or operators.

## Reviewers

- [x] @arbrandes
- [x] @bradenmacdonald 
- [ ] @doctoryes

## Supporting information and related

See the [changes to Content Libraries MFE](https://github.com/edx/frontend-app-library-authoring/pull/24).

## Testing instructions

1. Deploy an EdX devstack.
1. [Enable API access to the admin user](https://course-catalog-api-guide.readthedocs.io/en/latest/authentication/index.html).
1. Follow the [instruction to enable Content Libraries and Blockstore](https://github.com/edx/frontend-app-library-authoring/).
1. Log in to Studio, go to Content Libraries V2 UI (default devstack URL: http://localhost:3001).
1. Create a library, grab its "usage key" from the URL.

Then execute the following:

For API-based importing:

```
sudo -Hu edxapp bash
cd && . edxapp_env  && . ./venvs/edxapp/bin/activate && cd edx-platform/
OAUTH_KEY=
OAUTH_SECRET=
cat > oauth.creds <<EOF
${OAUTH_KEY)
$(OAUTH_SECRET)
EOF
python manage.py cms content_libraries_import lib:OC:cs course-v1:edX+DemoX+Demo_Course \
    api \
    --lms-url http://edx.devstack.lms:18000 --studio-url http://edx.devstack.studio:18010 --oauth-creds-file oauth.creds 
```

For local importing:

```
python manage.py cms content_libraries_import lib:OC:cs course-v1:edX+DemoX+Demo_Course modulestore
```

To test the APIs, please refer to [the MFE PR](https://github.com/edx/frontend-app-library-authoring/pull/24).

## Deadline

None.

## Other information

No dependencies or special concerns.

**Settings**:
```yaml
EDXAPP_FEATURES:
    ENABLE_LIBRARY_AUTHORING_MICROFRONTEND: true

EDXAPP_LIBRARY_AUTHORING_MICROFRONTEND_URL: "https://library-authoring.{{ EDXAPP_LMS_BASE }}"

TRUSTED_DOMAINS:
  - "library-authoring.{{ EDXAPP_LMS_BASE }}"

EDXAPP_ENABLE_CORS_HEADERS: yes
EDXAPP_ENABLE_CROSS_DOMAIN_CSRF_COOKIE: yes
EDXAPP_CORS_ORIGIN_WHITELIST: "{{ TRUSTED_DOMAINS }}"
EDXAPP_LOGIN_REDIRECT_WHITELIST: "{{ TRUSTED_DOMAINS }}"
EDXAPP_CSRF_TRUSTED_ORIGINS: "{{ TRUSTED_DOMAINS }}"

MFE_DEPLOY_STANDALONE_NGINX: yes

MFES:
  - name: library-authoring
    repo: frontend-app-library-authoring
    env_extra:
        STUDIO_BASE_URL: https://{{ EDXAPP_CMS_BASE }}
```

edx-webhook commented:
Thanks for the pull request, @jvdm! I've created [OSPR-5800](https://openedx.atlassian.net/browse/OSPR-5800) to keep track of it in JIRA, where we prioritize reviews. Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.




<!-- comment:external_pr -->
<!-- jenkins ok to test -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

natabene commented:
@jvdm Thank you for your contribution. Please let me know it is ready for our review.

jvdm commented:
> @jvdm Thank you for your contribution. Please let me know it is ready for our review.

It is ready.  Thanks.

natabene commented:
@sarina I thinks this might be with your team, could someone review it? If not, please point me in the right direction.

sarina commented:
Content libraries & blockstore belong to T&L, thanks!

natabene commented:
@sarina Indeed, thanks, sorry about that.

arch-bom-gocd-alerts commented:
📣 💥 Heads-up: You must either **rebase onto master** or **merge master into your branch** to avoid breaking the build.

We recently removed diff-quality and introduced lint-amnesty. This means that the automated quality check that has run on your branch doesn't work the same way it will on master. If you have introduced any quality failures, they might pass on the PR but then break the build on master.

This branch has been detected to not have commit 2e335653 as an ancestor. Here's how to see for yourself:

```
git merge-base --is-ancestor 2e335653 jvdm/libraries-v2-import-from-courseware && echo "You're all set" || echo "Please rebase onto master or merge master to your branch"
```

If you have any questions, please reach out to the Architecture team (either #edx-shared-architecture on Open edX Slack or #architecture on edX internal).

natabene commented:
@jvdm Could you look into the failing test?

doctoryes commented on openedx/core/djangoapps/content_libraries/management/commands/content_libraries_import.py:
```diff
@@ -0,0 +1,376 @@
+        elif options['method'] == 'modulestore':
+            edx_client = EdxModulestoreClient()
+        else:
+            assert False, f"Method not supported: {options['method']}"
```
Nit: `raise CommandError()` as above instead?

doctoryes commented on openedx/core/djangoapps/content_libraries/management/commands/content_libraries_import.py:
```diff
@@ -0,0 +1,376 @@
+        else:
+            assert False, f"Method not supported: {options['method']}"
+
+        # Query the course and rerieve all course blocks.
```
`retrieve`

doctoryes commented on openedx/core/djangoapps/content_libraries/management/commands/content_libraries_import.py:
```diff
@@ -0,0 +1,376 @@
+        export_keys = edx_client.get_export_keys(options['course-key'])
+        if not export_keys:
+            raise CommandError("The courseware course specified does not have "
+                               "any exportable content.  No action take.")
```
`No action taken.`

jvdm replied:
+1

doctoryes commented on openedx/core/djangoapps/content_libraries/management/commands/content_libraries_import.py:
```diff
@@ -0,0 +1,376 @@
+            files = [
+                f.path for f in
+                contentlib_api.get_library_block_static_asset_files(blockstore_key)
+            ]
```
It seems inefficient to get all static asset files each time through this loop - but I understand it makes for simpler code so 🤷🏼‍♀️ .

jvdm replied:
I am not aware of a bulk API to retrieve static files metadata from blockstore bundles. Indeed, that would be a great follow up for this for massive course imports (e.g. > 10K blocks). I am glad you seem inclined to agree that this can be done as a follow-up while keeping this first version simpler.

doctoryes commented on openedx/core/djangoapps/content_libraries/tests/test_command_content_libraries_import.py:
```diff
@@ -0,0 +1,106 @@
+"""
+Tests for managent command "importcourseware".
```
`management`. And the mgmt command is called `content_libraries_import`.

jvdm replied:
+1

doctoryes commented on openedx/core/djangoapps/content_libraries/tests/test_command_content_libraries_import.py:
```diff
@@ -0,0 +1,106 @@
+@mock.patch('openedx.core.djangoapps.content_libraries.management.commands.content_libraries_import.contentlib_api')
+class ImportCoursewareTest(TestCase):
+    """
+    Unit tests for importcourseware command.
```
`importcourseware` -> `content_libraries_import`

jvdm replied:
+1

doctoryes commented on openedx/core/djangoapps/content_libraries/tests/test_command_content_libraries_import.py:
```diff
@@ -0,0 +1,106 @@
+            self.call_command()
+
+    @ddt.data("drag-and-drop-v2", "problem", "html", "video")
+    def test_call_when_block_with_olx(self, block_type, api_mock, edx_class_mock):
```
`block_type` isn't used in the code below?

doctoryes replied:
But perhaps a test of a course with a non-exportable block_type is worth writing here?

doctoryes commented on openedx/core/djangoapps/content_libraries/management/commands/content_libraries_import.py:
```diff
@@ -0,0 +1,376 @@
+            resp = self._get(path)
+        else:
+            resp = requests.get(asset_file['url'])
+        resp.raise_for_status()
```
A test to exercise this code path is not present below.

jvdm replied:
Thank you for following up on the logic and capturing this. I've fixed it and added some test cases to cover the scenario.

doctoryes reviewed (commented):
Initial review complete.

jvdm commented on openedx/core/djangoapps/content_libraries/management/commands/content_libraries_import.py:
```diff
@@ -0,0 +1,376 @@
+
+from django.conf import settings
+from django.core.management import BaseCommand
+from django.core.management import CommandError
```
+1

jvdm commented on openedx/core/djangoapps/content_libraries/management/commands/content_libraries_import.py:
```diff
@@ -0,0 +1,376 @@
+from opaque_keys.edx.keys import CourseKey
+from opaque_keys.edx.keys import UsageKey
+from opaque_keys.edx.locator import LibraryLocatorV2
+from opaque_keys.edx.locator import LibraryUsageLocatorV2
```
+1

jvdm commented on openedx/core/djangoapps/content_libraries/management/commands/content_libraries_import.py:
```diff
@@ -0,0 +1,376 @@
+        """
+        Get the block's OLX and static files, if any.
+        """
+        raise NotImplementedError()
```
+1

jvdm commented on openedx/core/djangoapps/content_libraries/management/commands/content_libraries_import.py:
```diff
@@ -0,0 +1,376 @@
+
+    def write_error(self, *args, **kwds):
+        """
+        Write error messagses to stdout.
```
+1

arbrandes commented on openedx/core/djangoapps/olx_rest_api/adapters.py:
```diff
@@ -86,7 +86,7 @@ def collect_assets_from_text(text, course_id, include_content=False):
         info = {
             'path': path,
-            'url': '/' + str(course_id.make_asset_key("asset", path)),
+            'url': '/' + str(StaticContent.compute_location(course_id, path)),
```
Note to reviewer: this was necessary in order to fix the import of static files in subpaths.  (While importing from "static/img.jpg worked, "static/images/img.jpg" didn't.)

arbrandes commented:
@marcotuts, this is the backend component of the import-from-course-to-library feature of frontend-app-library-authoring, in case can also review it.

@bradenmacdonald, as core contributor and original author of a lot of related code, is this something you'd have time to review?

bradenmacdonald commented:
@arbrandes  Yes I'd be happy to review.

arbrandes commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1051,3 +1089,285 @@ def revert_changes(library_key):
+        # Import OLX and publish.
+
+        set_library_block_olx(blockstore_key, block_data['olx'])
+        publish_changes(blockstore_key.lib_key)
```
@jvdm, can you please pull this in?  https://github.com/open-craft/edx-platform/pull/379/commits/ee4c4c469a8600bd2a23812ae3a829c145cac847

For performance reasons, we can only publish at the end of the process.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/models.py:
```diff
@@ -127,3 +132,82 @@ def save(self, *args, **kwargs):  # lint-amnesty, pylint: disable=arguments-diff
+        self.save()
+
+    def __str__(self):
+        return f'{self.library} #{self.pk}'
```
Maybe include the course ID too?

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/views.py:
```diff
@@ -689,3 +696,130 @@ def delete(self, request, usage_key_str, file_path):
+@view_auth_classes()
+class LibraryImportableCourseView(APIView):
+    """
+    List libraries that can be imported by the given user.
```
This docstring and the `query_parameter` docs below are wrong, because this API is only about courses, not libraries.

In fact, I don't really think this belongs in the libraries REST API area because it has nothing to do with libraries. It only lists courses that the user has Studio access to.

I believe that our current LabXchange import code makes one call to the existing `/api/courses/v1/courses/` API to get all courses the user can read, then makes one call to `/api/enrollment/v1/` to get all courses that the user has Instructor access to, and then combines that data.

So I don't think that a new API is needed, but if you want to make it easier for API clients, can we just an add optional query parameter to the existing `/api/courses/v1/courses/` endpoint to allow filtering by role?

jvdm replied:
> So I don't think that a new API is needed, but if you want to make it easier for API clients, can we just add an optional query parameter to the existing /api/courses/v1/courses/ endpoint to allow filtering by role?

I totally agree here. But it is not immediately trivial to include the filter to `/api/courses/v1/courses`, and it would take some additional effort to get there. My suggestion is, let's go with the two API calls while we plan the work on expanding that API to accommodate this use case, since this is obviously the long-term solution.

We would have to update our MFE patch @gabor-boros. Any blockers against this approach?

gabor-boros replied:
@jvdm We thought about this approach with @arbrandes, though in the middle of the implementation of the frontend I proposed to have this endpoint to simplify API client calls and I found a blocker that time.

Since then, the frontend implementation has been changed, and this approach could be solved on the frontend if needed, however, I'm a bit afraid that the API calls would be pretty slow in that case.

bradenmacdonald replied:
@gabor-boros Why would those two API calls be significantly slower than this one? (I'm assuming you can make the two calls in parallel btw)

gabor-boros replied:
I cannot remember exactly how the courses API response fields are prepared on the backend. As far as I remember -- I may remember incorrectly -- the data returned by that API collected from several places that slows the processing of 50 course responses compared to this API.

Though I'd like to highlight again that I thought about the solution you proposed in the first place.

jvdm replied:
I think we can move on from [this comment from the reviewer](https://github.com/edx/edx-platform/pull/27715#issuecomment-895469980) with the current API and capture this as a follow-up. I can capture the words to add a comment to the patch to explain why we are pushing this. Let me know if someone disagrees.

@gabor-boros:

> the data returned by that API collected from several places that slows the processing of 50 course responses compared to this API.

How slow was it? Would it impact UX that much. I think the course import user flow is not exercised frequently by the same user, and maybe we are prematurely optimizing for performance but harnessing the separation of concerns?

gabor-boros replied:
@jvdm I have no measurements from real usage.

arbrandes replied:
@jvdm, @gabor-boros, have we considered pagination?  Wouldn't splitting the API call make paginating in the frontend unreliable?

bradenmacdonald replied:
@arbrandes @gabor-boros We already use the existing two APIs from the LabXchange app for this _exact_ same use case, and I believe it works fine. To handle the case of staff users who have access to hundreds of courses, the backend caches (per-user) the full list of courses that a user can see. The frontend can then get that data from the backend, paginated however it wants.

Code (not open source) is [here](https://gitlab.com/opencraft/client/LabXchange/labxchange-dev/-/blob/616c1543b96c4cc331f541eeab48654f98fcd6c6/backend/labxchange/service_clients/edxapp/client.py#L665-724).

In the new use case here, there is no separate backend to do caching but the frontend is presumably a SPA so it can just retrieve and cache the entire list of courses, then cross-reference it with the list of roles and display it paginated. In fact it's arguably better to do that on the frontend so you can display a progress bar / loading message while it loads.

arbrandes replied:
@bradenmacdonald, > it can just retrieve and cache the entire list of courses, then cross-reference it with the list of roles and display it paginated.

I am suspicious of doing frontend pagination based on two or more disparate, and potentially large, data sources.  Sure, it's doable, but it is complicated, second-guesses DRF, and is generally worse as far as frontend performance goes.  Sure, if Open edX gives us no other solution that's what we'll have to do, but I think we should at least aim towards porting the functionality to the right namespace.

arbrandes replied:
Addendum: the user-facing side supports searching, which would further the frontend (and pagination) further.

bradenmacdonald replied:
OK. My bottom line is that the endpoint in question is a type of course search/list API and it belongs in the same namespace as the other course search/list APIs, or perhaps even should be just a filter/parameter on one of those existing APIs. And because of that, I think it should at least be submitted as a separate PR so as not to hold up the rest of this course import functionality, and to make it easier for people who work on the course APIs to be aware of it and comment.

bradenmacdonald reviewed (changes requested):
Very nice work. Just some minor comments and requests.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1051,3 +1089,285 @@ def revert_changes(library_key):
+            usage_id = block.scope_ids.usage_id
+            if usage_id in export_keys:
+                continue
+            if usage_id.block_type in self.EXPORTABLE_BLOCK_TYPES:
```
Some additional context: currently this API imports _all_ "exportable" blocks from the course, but I don't think that will be a common workflow. Much more common will be that users request a list of all the blocks in the course, manually choose a subset of them (or select all of a specific type, like problem), and then import that subset. (I know that the latter workflow may not be fully implemented until later, but I think it is the most important.)

arbrandes replied:
@bradenmacdonald,

> I don't think that will be a common workflow.

I agree, but it is, in fact, the first workflow requested by a serious user that's not LabXchange.  (As a result, we ended up with 1500-block libraries and hit all sorts of Blockstore performance bottlenecks, but hey, at least we now know they're there. ;)

bradenmacdonald replied:
Ah, good to know :)

bradenmacdonald replied:
Just noting that I still don't see the rationale for having `EXPORTABLE_BLOCK_TYPES` nor for why it contains the values it currently does. Can we add an explanatory comment (why) and perhaps also a `# TODO: make this a command line option, including an option for "all block types"` ?

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1051,3 +1089,285 @@ def revert_changes(library_key):
+                block_type=modulestore_key.block_type,
+                usage_id=modulestore_key.block_id,
+            )
+            get_library_block(blockstore_key)
```
>  does it make sense to compose the Blockstore's block usage_id with the course_key something like: usage_id=hash(modulestore_key.course_key + modulestore_key.block_id)?

That's a solution that I've used before for similar problems, and I think it would work well, but it does add a layer of complexity/indirection that can make it harder to debug issues with content (because you can't easily tell from the resulting `block_id` which course it came from).

I would suggest that we either:
(1) Put some explicit comments in the code explaining that this is meant to be idempotent, and when the command is run, count the number of existing blocks and just print a message like ("17 blocks already existed in the destination library and were not imported."). We don't necessarily have to worry about or support the situation where users import from multiple courses to the same library, as long as we're clear about what we're doing and the potential issues for that use case.
or
(2) Do what you're suggesting but use `usage_id = modulestore_key.block_id + "_" + hash(modulestore_key.course_key)[0,3]` so that humans can look at the ID and tell what the original course-scoped block ID was.

jvdm replied:
I like your suggested approach. Deal.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/management/commands/content_libraries_import.py:
```diff
@@ -0,0 +1,156 @@
+            self.stdout.write(f"{block_num}/{block_count}: {block_key}: ", ending='')
+            if exception:
+                self.write_error('❌')
+                self.stderr.write(f"Failed to import modulestore block: {exception}")
```
Is that a common convention? (and if so, why is the functionality not already baked into the superclass here?) If so, then we're all good. If not super common, please just include a comment explaining it for the dummies like me :)

jvdm replied:
> Is that a common convention?

Talking about the *nix command line in general? I've seen it many times, although I am not entitled to say, "this is a common convention around all command lines in the world". I guess it would be a point for style and discussion. I've seen many commands regard `stderr` as a place to log errors and diagnostic-related information, while `stdout` is used to show expected output or summarized progress. One usage pattern is: You call the command with `foobar 2>errors.log`, get a cleaner output, easier to see the big picture, and if something went wrong and you want to dive into the details of the errors, you check stderr. It is advantageous when your command will be used in scripts or automated environments. BTW, this is what I mean when I said "`2>/dev/null`-style calls", although redirecting to `/dev/null`, you are pretty much ignoring the details. My point was, you can redirect stderr to another place other than stdout.

> why is the functionality not already baked into the superclass here?

It is. They allow you to write to different output streams and allows you to pretty-print the output as well. I just added two wrappers called `write_error(...)` and `write_success(...)` to colorize the output according to upstream's convention of what would be a good color for error and success. I could just have used `self.stdout.write(self.style.ERROR(...))` instead. I think it is a little bit longer and repetitive, hence the wrapper method.

> If not super common, please just include a comment explaining it

I'll get rid of the helper method, which will make it explicit and less confusing.

bradenmacdonald commented:
@jvdm Please address the test failure and any remaining comments that you'd like to do now, and ping me when this is ready for final rest + approval. If you want to defer addressing some of my comments until future, it's fine with me, as long as there is a relevant comment in the codebase itself for explanation/context/tracking.

jvdm commented:
@bradenmacdonald:

> @jvdm Please address the test failure and any remaining comments that you'd like to do now, and ping me when this is ready for final test + approval.

I've followed up on the pending comments, ready for final testing and approval.

The remaining points that are outstanding are:

* TODO Making the command more efficient with bulk imports
* TODO Change MFE UI to consume better import course API

I am capturing those as a follow to this PR.

bradenmacdonald commented:
Began testing this today but got held up by devstack issues, sorry for the delayed approval.

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1051,3 +1089,295 @@ def revert_changes(library_key):
+        # course key into the imported block.
+
+        block_data = self.get_block_data(modulestore_key)
+        course_key_hash = str(abs(hash(str(modulestore_key.course_key))))
```
During testing today, I found an issue with this: the python `hash()` function is not a stable hash. So if you run the import command twice, it will created duplicate blocks in the library.

```
$ python -c 'print(str(abs(hash(str("course-v1:edX+DemoX+Demo_Course")))))'
7449256757677617019
$ python -c 'print(str(abs(hash(str("course-v1:edX+DemoX+Demo_Course")))))'
3804951377088993027
```

Please switch to a proper hash from `hashlib` like BLAKE2.

Second, IMHO these hashes are a bit too long. Using a real example from the edX demo course, one imported block got this ID:

```
lb:DevX:IMPORTED:problem:python_grader_3595429935507010447
```

I think a short 3-5 digit alphanumeric hash (or 7 like git uses for commit short IDs) would be more than sufficient for virtually eliminating conflicts between courses while still being human-friendly:

```
lb:DevX:IMPORTED:problem:python_grader_f8h39
```
But feel free to ignore this second part of the feedback; it's not that important.

jvdm replied:
FYI, here's an example of the resulting usage key with the latest changes:

```
lb:OC:clean:html:b8f31cd5bb6a74c9888c478ec72a75ffc_cdwdim25pmeavdvqv
```

bradenmacdonald reviewed (changes requested):
Looking great! I just tested this and am conditionally approving if you make two changes:

1. fix the hash function issue I pointed out.
2. Move the `importable_blocks` REST API to a separate PR for now.

- [x] I tested this: testing both a modulestore and an oauth-based import of the demo course to a library
- [x] I read through the code
- [x] I checked for accessibility issues: n/a, no UI
- [x] Includes documentation - comments and docstrings yes

jvdm commented:
> 2\. Move the `importable_blocks` REST API to a separate PR for now.

If we don't ship the API, the MFE won't be able to onboard the backend. So what is the plan moving forward?

bradenmacdonald commented:
@jvdm

> If we don't ship the API, the MFE won't be able to onboard the backend. So what is the plan moving forward?

I am assuming that there is no immediate need to use the MFE with `courses.edx.org` as its backend in the coming week; if that is the case, let me know.

I was thinking that for now, you'd just move the existing API endpoint code to a separate branch/PR so that the dev team working on this project can continue using it to test the MFE, for local development, etc. And in then in that PR we can discuss if that's the right approach (I don't think it is) or figure out a new approach. But basically I think it's better to maintain that temporary endpoint as a branch rather than merging it into the codebase with the intention of soon changing or removing it.

As for what the ultimate solution should be, I think that it would be one of (A) modify the existing course list API to return the data we need, when the right filter/parameters are passed, (B) modify the frontend to just use the existing two APIs (course list + get course roles), or (C) show that this new API is something distinct and necessary, and add it to the platform (but in the course API namespace, not the library API namespace, and in a separate PR with a relevant title so that people watching for course-related API PRs will see it).

jvdm commented:
@bradenmacdonald:

> I was thinking that for now, you'd just move the existing API endpoint code to a separate branch/PR so that the dev team working on this project can continue using it to test the MFE, for local development, etc. And in then in that PR we can discuss if that's the right approach (I don't think it is) or figure out a new approach. But basically I think it's better to maintain that temporary endpoint as a branch rather than merging it into the codebase with the intention of soon changing or removing it.

OK, thanks for clarifying the next steps. I'll make the updates to the PR and ping you after that.

jvdm commented:
> @bradenmacdonald:
> 
> > I was thinking that for now, you'd just move the existing API endpoint code to a separate branch/PR so that the dev team working on this project can continue using it to test the MFE, for local development, etc. And in then in that PR we can discuss if that's the right approach (I don't think it is) or figure out a new approach. But basically I think it's better to maintain that temporary endpoint as a branch rather than merging it into the codebase with the intention of soon changing or removing it.
> 
> OK, thanks for clarifying the next steps. I'll make the updates to the PR and ping you after that.

Done. BTW, here's the (draft) PR with the importable courses API: https://github.com/edx/edx-platform/pull/28464

bradenmacdonald reviewed (approved):
Thanks, good to merge now! Just needs to be squashed down to 1-2 commits, although I can do that as I merge if you like.

I will merge this tomorrow AM.

jvdm commented:
> Thanks, good to merge now! Just needs to be squashed down to 1-2 commits, although I can do that as I merge if you like.
> 
> I will merge this tomorrow AM.

Rebased onto `origin/master` and squashed.
